internal/deb/extract.go

```diff
@@ -162,81 +200,151 @@ func extractData(dataReader io.Reader, options *ExtractOptions) error {
 			continue
 		}
 		sourcePath = sourcePath[1:]
-		globPath, ok := shouldExtract(sourcePath)
-		if !ok {
-			continue
+		sourceMode := tarHeader.FileInfo().Mode()
+
+		globPath := ""
+		extractInfos := options.Extract[sourcePath]
+
+		if extractInfos == nil {
+			for _, glob := range globs {
+				if strdist.GlobPath(glob, sourcePath) {
+					globPath = glob
+					extractInfos = []ExtractInfo{{Path: glob}}
+					break
+				}
+			}
 		}
 
-		sourceIsDir := sourcePath[len(sourcePath)-1] == '/'
+		// Is this a directory path that was not requested?
+		if extractInfos == nil && sourceMode.IsDir() {
+			if info := dirInfos[sourcePath]; info.mode != sourceMode {
+				// We have not seen this directory yet, or we
+				// have seen or created it with a different mode
+				// before. Record the source path mode.
+				info.mode = sourceMode
+				dirInfos[sourcePath] = info
+				if info.created {
+					// We have created this directory before
+					// with a different mode. Create it
+					// again with the proper mode.
+					extractInfos = []ExtractInfo{{Path: sourcePath}}
+				}
+			}
+		}
 
-		//debugf("Extracting header: %#v", tarHeader)
+		if extractInfos == nil {
+			continue
+		}
 
-		var extractInfos []ExtractInfo
 		if globPath != "" {
-			extractInfos = options.Extract[globPath]
-			delete(pendingPaths, globPath)
 			if options.Globbed != nil {
 				options.Globbed[globPath] = append(options.Globbed[globPath], sourcePath)
 			}
+			delete(pendingPaths, globPath)
 		} else {
-			extractInfos, ok = options.Extract[sourcePath]
-			if ok {
-				delete(pendingPaths, sourcePath)
-			} else {
-				// Base directory for extracted content. Relevant mainly to preserve
-				// the metadata, since the extracted content itself will also create
-				// any missing directories unaccounted for in the options.
-				err := fsutil.Create(&fsutil.CreateOptions{
-					Path:        filepath.Join(options.TargetDir, sourcePath),
-					Mode:        tarHeader.FileInfo().Mode(),
-					MakeParents: true,
-				})
-				if err != nil {
+			delete(pendingPaths, sourcePath)
+		}
+
+		// createParents creates missing parent directories of the path
+		// with modes with which they were encountered in the tarball or
+		// 0755 if they were not encountered yet.
+		var createParents func(path string) error
+		createParents = func(path string) error {
```

Comment:
We'll need to take a step back on these changes and understand what we want to do here, and what's the best path for it. The diff here is clearly making something that was already complex and making it even less great.

Reply:
This is not supposed to be part of this PR, so I'm closing it here as the PR message (from commit message), doesn't give enough context. The PR introducing this change is here: https://github.com/canonical/chisel/pull/74

Reply:
@woky can you please add the chain of dependencies to this PR's description so that we know in which order to review? #74 is one of them. What else?

Reply:
Done. Nothing else except #74.